lib/utils/deprecation/deprecator.rb

```diff
@@ -0,0 +1,109 @@
+
+require 'utils/deprecation/config_file'
+require 'inspec/log'
+
+module Inspec
+  module Deprecation
+    class Deprecator
+      attr_reader :config, :groups
+
+      # This is used only in functional testing
+      def self.class_test_cfg_io(io)
+        @@test_cfg_io = io # rubocop: disable Style/ClassVars
+      end
+
+      def initialize(opts = {})
+        @@test_cfg_io ||= nil # rubocop: disable Style/ClassVars
+        @config = Inspec::Deprecation::ConfigFile.new(opts[:config_io] || @@test_cfg_io)
+        @groups = @config.groups
+      end
+
+      def handle_deprecation(group_name, message, opts = {})
+        group = groups[group_name.to_sym] # Will be nil if unrecognized group
+        annotate_stack_information(opts)
+        assembled_message = assemble_message(message, group, group_name, opts)
+
+        action = (group ? group[:action] : :warn) || :warn
+        action_method = ('handle_' + action.to_s + '_action').to_sym
+        send(action_method, assembled_message, group)
+      end
+
+      private
+
+      def annotate_stack_information(opts)
```

Comment:
How do you feel about stack-probing heuristics?

Reply:
brittle, but necessary.  the client does this in its error message decorator and in the deprecations it throws.

Reply:
I've added code to filter out RSpec, as suggested.